header-rewrite-filter/header_processor.cc

```diff
@@ -53,32 +53,32 @@ namespace HeaderRewriteFilter {
         return absl::OkStatus();
     }
 
-    absl::Status HeaderProcessor::evaluateCondition(SetBoolProcessorMapSharedPtr bool_processors) {
+    std::pair<absl::Status, bool> HeaderProcessor::evaluateCondition(SetBoolProcessorMapSharedPtr bool_processors) {
         // call ConditionProcessor executeOperation; if it is null, return true
         bool result = true;
         ConditionProcessorSharedPtr condition_processor = getConditionProcessor();
         if (condition_processor) {
-            const absl::Status status = condition_processor->executeOperation(bool_processors);
+            const std::pair<absl::Status, bool> condition = condition_processor->executeOperation(bool_processors);
+            const absl::Status status = std::get<0>(condition);
             if (status != absl::OkStatus()) {
-                return status;
+                return std::make_pair(status, false);
             }
-            result = condition_processor->getConditionResult();
+            result = std::get<1>(condition);
         }
-        setCondition(result);
 
-        return absl::OkStatus();
+        return std::make_pair(absl::OkStatus(), result);
```

Comment:
Nit: I think this would be a little safer if it were restructured to not have a `result` variable, and also to return immediately when it knows it can:
```
        ConditionProcessorSharedPtr condition_processor = getConditionProcessor();
        if (!condition_processor) {
                return std::make_pair(absl::OkStatus(), true);
        }
        const std::pair<absl::Status, bool> condition = condition_processor->executeOperation(bool_processors);
        const absl::Status status = std::get<0>(condition);
        if (status != absl::OkStatus()) {
            return std::make_pair(status, false);
        }
        return std::get<1>(condition);
```

**Long Version**

So, there are two schools of thought here. Some people prefer to never return early, and always fully trace out every condition branch, and only ever return at the end of a function.

Benefits:
- if you mentally follow it, you can visibly see every branch and possibility
- theoretically no possible state gets accidentally missed
Disadvantages:
- it leads to deep nesting, which
    - generally means more Cyclomatic Complexity, which is bad for a lot of reasons
    - is generally hard to read
- it means you have to keep a lot more in your head

The other school of thought is to "return as soon as you can"
Benefits:
- much less nesting
- generally easier to read
    - with a certain mindset, you have to mentally say "ok, this condition is always X from now on"
Disadvantages:
- have to mentally track what conditions are always true after the returns
    - Which is usually always either the "good" or "error" path, so _usually_ easier

The first used to be common, but I think the second is more widely preferred these days. But I know good programmers who prefer the first.

-- 

The second thing here is Mutable State. I've kind of alluded to this in previous comments. IMO Mutable State is an enormous source of bugs. Figuring out the global state of your program at any given time, when something goes wrong, is a debugging nightmare.

Not only in individual functions like here, but in the larger structure of a program: 
Whether you track the state of your program by setting an object variable, then calling functions that later read that variable; 
or whether you call functions and pass variables that never change (immutable).

In both cases, you're tracking the state of the world. But in the second, it's implicit, and immutable. At any point in the execution, it's much easier to see, just by looking at the code, what the state of the world is, and how it got there. 

If you're using a debugger, you can know exactly what happened simply by looking at the stack; whereas with Mutable State, you can see the variables with a debugger (or prints/logs), but you likely have no idea how they got the values they have.

"Stateful" programming also used to be widespread (probably because it models the physical computer), "Immutability" came from Functional Programming, and for a long time was a minority, but lately I think it's also becoming widely accepted as better/safer/easier to read/easier to debug/etc.

- To be clear, some state is necessary, like your parsed Processors here, or e.g. in a game you'd need to track the state of the world, or an app that gathers metrics and keeps a running sum. Immutability is about not using Mutable State in the _flow_ of your program, not the changing state of the world, if that makes sense (though that's also possible, it's pretty extreme)

--

Anyway, I'm just trying to be helpful, but if you already know this stuff or it comes off as Mansplaining or anything, feel free to ignore my long-winded explanations, or let me know and I can put my Teacher Hat on less often 😂

Reply:
Thanks for the suggestion, Rob!

Your explanations are always super helpful for me to learn better C++ :) Thank you so much for taking the time and effort to provide them, I really appreciate it!